// fastify-i18next/lib/languageLookups/cookie.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _cookies = _interopRequireDefault(require("cookies"));

var _default = {
  name: 'cookie',
  lookup: function lookup(req, res, options) {
    var found;

    if (options.lookupCookie && typeof req !== 'undefined') {
      if (req.cookies) {
        found = req.cookies[options.lookupCookie];
      } else {
        var cookies = new _cookies["default"](req, res);
        found = cookies.get(options.lookupCookie);
      }
    }

    return found;
  },
  cacheUserLanguage: function cacheUserLanguage(req, res, lng) {
    var options = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};

    if (options.lookupCookie && req !== 'undefined' && !(res._headerSent || res.headersSent)) {
      var cookies = new _cookies["default"](req, res);
      var expirationDate = options.cookieExpirationDate;

      if (!expirationDate) {
        expirationDate = new Date();
        expirationDate.setFullYear(expirationDate.getFullYear() + 1);
      }

      var cookieOptions = {
        expires: expirationDate,
        domain: options.cookieDomain,
        httpOnly: false,
        overwrite: true
      };

      if (options.hasOwnProperty('cookieSecure')) {
        cookieOptions.secure = options.cookieSecure;
      }

      cookies.set(options.lookupCookie, lng, cookieOptions);
    }
  }
};
exports["default"] = _default;
